### ISSUE TITLE
Following up on Issue 4248  Still some possible problems when using multiple instances..

### ISSUE THREAD

**Issue by todd-prior** [darktable-org/darktable#4280]
**Describe the bug**
Using a recent build I made some additional edits to images that were already edited. These were JPG images. The additional edits introduced duplicate instances. The edits appeared fine when I exited DT but reopening the next day the images were scrambled. I have DT set to look for new xmp files. There seems to be some processing that went on. There are additional lines in the history that don't match up to the modules shown as activated when you compare. The edits are fine up to the point of the duplicate instance and after that they are scrambled. I will include the original images and the xmp before (I have versioning in place) the added edits and then the one that causes the issues. I will do this for a few examples...I have seen it now with color zones, denoise profiled and rgb curve and rgb levels modules. 

**To Reproduce**
The error occurs when reopening an image in DT that has had edits that introduce duplicate instances. Upon being reopened the images have additional entries in the history stack and these do not line up with the list of modules shown as active in Darkroom view. See the active modules vs the history in the screenshots...

**Expected behavior**
additional edits are not altered by any recalculation or reprocessing of the pixelpipe

**Screenshots**

![image](https://user-images.githubusercontent.com/44409419/74377355-be91d580-4db1-11ea-8451-967457f46524.png)
![image](https://user-images.githubusercontent.com/44409419/74377391-ce111e80-4db1-11ea-9ff8-5a0ae78f10fa.png)

![image](https://user-images.githubusercontent.com/44409419/74377432-dff2c180-4db1-11ea-834d-8e03169f7cc5.png)

Note this time no color zones 1 module listed in the active module but it is in the history
![image](https://user-images.githubusercontent.com/44409419/74377557-234d3000-4db2-11ea-895f-09edc456accf.png)

**Platform (please complete the following information):**
 Darktable Version: 3.1.0+606
 Windows 10  32 GB RAM 
 OpenCL activated 
 Geforce GT710 2GB  441.87 

**Additional context**
Zip file added to contain original image with xmp before and after the issue appears  

[Test images.zip](https://github.com/darktable-org/darktable/files/4195175/Test.images.zip)

**Comment by todd-prior** [comment]
As a follow up i will try to recreate the edit and if it looks fine I will capture the xmp before closing DT and then see if reopening changes anything..

**Comment by TurboGit** [comment]
@todd-prior : thanks for the effort, indeed any step by step procedure will help fix this issue. at the moment I didn't had such issue on my side.

**Comment by todd-prior** [comment]
@TurboGit I'm wondering if I created it by looking at the images on two different computers running slightly different versions.. I try to keep them in sync but maybe this is it....I had some shots I think I edited on 3.1.578 and might have viewed them at work on 3.1.438???  Just guessing. Basically they were underwater shots and while not likely best practice i was getting good visual results using levels twice ...once on unlinked channels and auto fitting them and then a second one on the linked rgb histogram to adjust the resulting white black and mid points.....so this was what most of the ones that went south were ie two instances of levels. It seems like this was translated at some point to the disconnect between the xmp and the modules shown in the panel and the corrupt image....I have worked around it by copying the stack to a new virgin duplicate and omitting the repeated module. This seems to be a way to fix it. Funny though for the image with two instances of profiled denoise i am sure i had only one instance. It seems like sometimes now and I don't remember this in the past that selecting the module creates a history entry and then adjusting the parameters creates a second one right after it. I wonder if it was a translation of the two history entries of that type that created the "new" instance??  In any case I will try to keep my eye out for any other cases of this. It seemed to happen on a folder of 200 or so images that I had been editing. I think maybe I looked at them at work on a slightly earlier version and so the xmp files may have gotten modified?? I have not seen it in any of my other folders yet but if I do I will do some more checking to try and see what conditions lead to this mismatch....very funny when I copy the stack and paste back deleting the double instance of the module and then add it back it gets labelled as for example rgb curve 2 with priority 2 and the image is fine with that ??  This really leads me to believe something was altered going back a few commits between the computers...I'll let you know if it is more wide spread but I have found a way to work around it I just wanted to report it in case it was an issue with all the coding done recently to try to reel in the order of iop and history and styles etc......thanks for your feedback

**Comment by todd-prior** [comment]
Did just find one in another folder...seems like multiple instances of denoise profiled were created??

![image](https://user-images.githubusercontent.com/44409419/74403079-118c7c80-4df4-11ea-80f8-2af7e8e151e3.png)

[20190706_164036_DT_01.jpg.xmp.txt](https://github.com/darktable-org/darktable/files/4196553/20190706_164036_DT_01.jpg.xmp.txt)

**Comment by TurboGit** [comment]
Ok, the XMP is wrong.

xmp_version         = 4 OK
iop_order_version = 2 OK (means new iop-order v3.0)

You have indeed multiple instances. So we should have in the XMP:

darktable:iop_order_list="....."

Can you double check your version of dt? What is the sha-1 you have  compiled ? I'm asking this as the fix for this issue has been introduce in **release-3.1.0-583-gca4db3f1b**.

Thanks.

**Comment by TurboGit** [comment]
> I had some shots I think I edited on 3.1.578

That's probably it then. Indeed this version does not contain the fix. You probably want to close this issue :) Make sure you have the latest dt version on all computers.

**Comment by Nilvus** [comment]
> > I had some shots I think I edited on 3.1.578
> 
> That's probably it then. Indeed this version does not contain the fix. You probably want to close this issue :) Make sure you have the latest dt version on all computers.

Sorry but as I have said on prévious issue from @todd-prior, I have had the same this week (so with the fix you made) with this issue, and each time all with 2 instances of denoise profiled module. One with a default non-local means auto mode and the second for new Y0U0V0 mode. This issue appear each time when copying a newly edited one with these 2 instances on an old image so with an old IOP order (all target images I have seen that were edited with darktable 2.6).

So it's too early to close this issue. Maybe it's related to copy/paste order module not updated (I precise that changing module order on these images doesn't fix the issue). I don't have test that with your open PR about copy/paste module order (maybe it's related to my open issue about that) and I don't know if using your darktable branch of that last PR include the fix. Hope that's helps.

**Comment by TurboGit** [comment]
@Nilvus : Ok, we can keep this one or open a new one. This one is clearly a wrong use and I do prefer keeping issues separated.

That being said I have tried again to reproduce without success! (I'm using current master - **release-3.1.0-639-g832c76283**).

Copied 2 denoise profile instances from a v3.0 order to a legacy order having denoise profile already or not. Nothing wrong, the edits are correct and even after restarting dt. I'm in the complete fog about this issue, so I'm waiting for a step by step reproducer.

**Comment by TurboGit** [comment]
Fount it, it is when you paste on the currently opened image. I was always copying on the currently open image and pasting into another one!

**Comment by todd-prior** [comment]
@TurboGit FYI these are roughly the versions i have been using and the dates...but my issues have been created I think during an update of the xmp with no copying involved....   
![image](https://user-images.githubusercontent.com/44409419/74449578-bfc31100-4e4a-11ea-8dcf-c08ec04626f5.png)

**Comment by TurboGit** [comment]
Should be fixed now! I won't dare to say that this is the last iop-order issue :)

**Comment by todd-prior** [comment]
Nope there is a bug in the new iop module…will post shortly

**Comment by todd-prior** [comment]
So this is it Pascal for sure . I found it during my trouble shooting but basically if you have an image with a iop order of 2.0 and you change it to 1.0 and then move to a new image when you come back it is scrambles….in the xmp the version is still 2.0 not changed to 1.0 and the iop_order_list line is gone….

I found this out by try to save an edit…so went back to my edit that was scrambled and copied the history stack back to a virgin duplicate without the duplicate modules and then I just readded them. This was all fine no image corruption and things looked fine. Likely because the duplicate was now version 2 and had the iop order list line. But if I took that image and changed the order to legacy I could see and update and the difference so still not scrambled but moving to a new image and coming back to it scrambled it…..looking in the xmp it was version 2 and no iop order so the corruption is in the xmp and I think the database must be fine as the image corruption does not show when you toggle back and forth only once the xmp writes…..Also doing the reverse does not restore the image as I suspect when it is already indicating version 2.0 iop it is not going to write a new iop order line…..so I think there is just a small bug in the translation between legacy and 3.0 order when written to the XMP. I am on Windows 10 3.1.0+606 currently …..

Hope this makes sense. If you need files let me know but I think this is the culprit…of course I am not sure about the code at all or the actions of this parameter it seems to be the issue…..

**Comment by todd-prior** [comment]
What I don't know is perhaps this only scrambles some images depending on what modules are used etc ...certainly for me it was ones that had multiple instances. I know that was experimenting on several image to compare how they looked with the different order and on several I preferred the legacy order and so that is how I ended up with the corrupt xmp files....I'll rebuild and try out your fix......for my corrupt edits if I add back an iop order line or manually edit to version 1 do you think they will behave..I may try it...all the values in that string seem to be zero but I really don't know fully what it does wrt to the iop order entry in the xmp??

**Comment by todd-prior** [comment]
Ya but in the end I still seem to have some files that are corrupted and to me the xmp looks okay……should both version 1 and version 2 iop order xmp files have the order list line or just version 2??

Because not all my corrupt files have the mismatch issue of being version 2 without this line. I have some that are strait up version 1 files but they do not seem to play well with current builds of DT at least up to 606 I will try to update today and see how it works…. But basically the problem starts at the second instance of the levels in some of my edits and not sure why it cannot be read??

**Comment by TurboGit** [comment]
@todd-prior : the iop-order list in the XMP should be there only if the order is custom or if there is multiple instances. For plain legacy or v3.0 order the list is not dumped.

**Comment by todd-prior** [comment]
Okay so that is my issue in the bad files I have no iop-order-list…..I will see if I can add that back I tried once but it did not work so I will play with the syntax of it…

**Comment by TurboGit** [comment]
You may try to move up the iop-instance 1 and back down to its position. It should (not sure) generate the needed iop-list.

**Comment by todd-prior** [comment]
I’ll have to do it manually but I think it works if I add it and get the coefficient correct I was not sure if it was the name or module it was referring to in the list.

I would try your suggestion but the issue is that there is no module to move. Its like the error occurs at the point where the xmp refers to  the next instance but it fails to go further so all modules after do not show. Basically it is as if you have moved to that point in your history stack and only the modules show that are active up to that point…so then even though there may be 5 or 6 more modules beyond this listed in the history stack they are not showing in the darkroom module tabs.

I found that if I compressed the history stack that removed some of the extra entries. The xmp file would at time seem to create 2 entries for each instance so you would have say 3 instances of rgb curve and there would be 6 entries….rgb curve rgb curve rgb curve 1 rgb curve1 rgb curve 2 rgb curve 2. It look like one was added for creating the instance and a second was created when you made any change in that instance….So compressing left the sequence with just one line for each of the 3 instances of the rgb curve module and then adding back the iop_order _list with with 3 entries 0, 1, and 2 seems to restore things…I don’t think is a massive number of edits but at least I can see how to fix it. I will test out the new fix you inserted today when I am home tonight and confirm that it corrects the missing iop order issue in the xmp.

Thanks for all the updates
